File: web/src/lib/nickname.test.ts
import { describe, it, expect } from "vitest";
import { randomNickname, validateNickname } from "./nickname";

describe("randomNickname", () => {
  it("generates 用户 + 4 digits in zh", () => {
    expect(randomNickname("zh")).toMatch(/^用户\d{4}$/);
  });

  it("generates User + 4 digits in en", () => {
    expect(randomNickname("en")).toMatch(/^User\d{4}$/);
  });
});

describe("validateNickname", () => {
  it("accepts Han, Latin, digits and underscore in 2–16 characters", () => {
    for (const ok of ["用户1234", "User", "_name_", "张三三", "aaaaaaaaaaaaaaaa", "字a1_"]) {
      expect(validateNickname(ok)).toBe("ok");
    }
  });

  it("rejects wrong length and disallowed characters", () => {
    for (const bad of ["a", "aaaaaaaaaaaaaaaaa", "用户-1", "用户 1", "🙂🙂"]) {
      expect(validateNickname(bad)).toBe("format");
    }
  });

  it("counts code points not bytes (16 CJK chars is legal)", () => {
    expect(validateNickname("布丁盒子布丁盒子布丁盒子布丁盒子")).toBe("ok");
  });
});

File: internal/productstate/nickname_test.go
package productstate

import "testing"

func TestValidateNickname(t *testing.T) {
	// Legal: Han, Latin, digits, underscore, mixed, boundary lengths.
	for _, ok := range []string{
		"用户1234",
		"User",
		"_name_",
		"张三三",
		"aaaaaaaaaaaaaaaa", // exactly 16
		"字a1_",
	} {
		if err := ValidateNickname(ok); err != nil {
			t.Errorf("ValidateNickname(%q) = %v, want nil", ok, err)
		}
	}

	// Illegal lengths and characters.
	for _, bad := range []string{
		"a",                 // 1 rune, below min
		"aaaaaaaaaaaaaaaaa", // 17 runes, above max
		"用户-1",              // hyphen not in set
		"用户 1",              // space not in set
		"🙂🙂",                // emoji not in set
	} {
		if err := ValidateNickname(bad); err == nil {
			t.Errorf("ValidateNickname(%q) = nil, want error", bad)
		}
	}
}

func TestValidateNicknameCountsRunesNotBytes(t *testing.T) {
	// 16 CJK runes is 48 bytes; byte-counting would wrongly reject it.
	if err := ValidateNickname("布丁盒子布丁盒子布丁盒子布丁盒子"); err != nil {
		t.Errorf("16-rune CJK nickname = %v, want nil", err)
	}
}

func TestSensitiveDefaultFailsClosed(t *testing.T) {
	if !Sensitive("anything") {
		t.Fatal("unwired sensitive gate must reject nicknames")
	}
}

func TestSensitiveIsSwappable(t *testing.T) {
	orig := Sensitive
	defer func() { Sensitive = orig }()
	Sensitive = func(v string) bool { return v == "敏感词" }
	if !Sensitive("敏感词") {
		t.Fatal("swapped Sensitive did not run")
	}
	if Sensitive("正常") {
		t.Fatal("swapped Sensitive over-matched")
	}
}
